File: packages/llm/src/messaging/templates.py
"""Template definitions and LLM prompts for each message variant."""

from __future__ import annotations

from .models import MessageVariant

# --- System prompt shared by all variants ---

SYSTEM_PROMPT = """\
Du schreibst eine kurze, persönliche Nachricht an jemanden, der eine Immobilie auf \
Kleinanzeigen inseriert hat. Du hast die Anzeige gelesen und etwas Bestimmtes ist \
dir aufgefallen.

REGELN:
- Maximal 80 Wörter, idealerweise 50-70
- Du bist KEIN Makler und bietest NICHTS an
- Du bist eine Person mit Immobilienerfahrung, die etwas Interessantes bemerkt hat
- Beginne mit einer konkreten Beobachtung über die Immobilie (NICHT "Hallo" oder "Sehr geehrte/r")
- Genau EIN Call-to-Action: eine einfache Frage, die leicht zu beantworten ist
- Kein kommerzieller Ton, keine Verkaufssprache
- Kein "ich könnte Ihnen helfen", "Zusammenarbeit", "Angebot", "kostenlos"
- Kein "Makler", "Vermittlung", "Provision", "Vermarktung"
- Natürlich, wie eine WhatsApp-Nachricht an einen Bekannten
- KEIN Ausrufezeichen am Anfang, maximal 1 im gesamten Text
- Die Nachricht muss mit genau einer Frage enden (?)
- Schreibe KEINE Grußformel am Ende (kein "Viele Grüße", "LG", "MfG")
- Beginne NICHT mit "Ich" oder "Mein"

{tone_instruction}

{variant_instruction}
"""

TONE_DU = "Duzen — schreibe informell mit 'du/ihr/euch'. Locker und freundlich."
TONE_SIE = "Siezen — schreibe mit 'Sie/Ihnen/Ihr'. Respektvoll aber nicht steif."


# --- Variant-specific instructions ---

VARIANT_INSTRUCTIONS: dict[MessageVariant, str] = {
    MessageVariant.SPECIFIC_OBSERVER: """\
STRATEGIE: "Der aufmerksame Beobachter"
Psychologie: Reziprozität + Sympathie

Deine Nachricht soll zeigen, dass du die Anzeige WIRKLICH gelesen hast. Führe mit \
einem hyper-spezifischen Detail, das dir aufgefallen ist. Zeige echte Neugier für \
etwas, worauf der Verkäufer stolz ist.

STRUKTUR:
1. Konkrete Beobachtung über ein besonderes Merkmal der Immobilie
2. Warum das deine Aufmerksamkeit geweckt hat (1 Satz)
3. Eine leichte, persönliche Frage dazu

Die Frage soll etwas sein, worüber der Verkäufer GERNE redet — z.B. die Geschichte \
hinter einer Renovation, wie sie auf die Idee für den Wintergarten kamen, etc.""",

    MessageVariant.MARKET_INSIDER: """\
STRATEGIE: "Der Marktkenner"
Psychologie: Autorität + Neugier

Zeige Marktkenntnis durch eine spezifische Preis-Beobachtung. Erstelle eine \
Wissenslücke, die der Verkäufer füllen will.

STRUKTUR:
1. Preis-pro-qm Beobachtung im Vergleich zum regionalen Markt
2. Eine unvollständige Andeutung ("das kann gewollt sein, aber...")
3. Offene Frage zur Preisfindung

WICHTIG: Sei NICHT belehrend oder herablassend. Formuliere es als Beobachtung, \
nicht als Kritik. Der Verkäufer soll neugierig werden, nicht defensiv.""",

    MessageVariant.EMPATHETIC_PEER: """\
STRATEGIE: "Der verständnisvolle Gleichgesinnte"
Psychologie: Social Proof + Sympathie

Positioniere dich als jemand, der selbst privat verkauft hat und die Herausforderungen \
kennt. Baue Rapport durch geteilte Erfahrung auf.

STRUKTUR:
1. Kurze empathische Aussage über ihre Situation (privat verkaufen ist aufwändig)
2. Minimale eigene Erfahrung teilen (1 Satz, glaubwürdig)
3. Frage nach ihren bisherigen Erfahrungen

Die Frage soll dem Verkäufer erlauben, Frust abzulassen — viele private Verkäufer \
sind genervt von Lowball-Angeboten und No-Shows.""",

    MessageVariant.CURIOUS_NEIGHBOR: """\
STRATEGIE: "Der neugierige Nachbar"
Psychologie: Sympathie + Knappheit

Positioniere dich als lokal verbundene Person, die das Inserat entdeckt hat. \
Betone die Seltenheit dieser Art von Objekt in der Gegend.

STRUKTUR:
1. Lokaler Bezug oder Beobachtung
2. Hinweis auf Knappheit/Besonderheit in der Region
3. Beiläufige, nachbarschaftliche Frage

Ton: So als würdest du einen Nachbarn am Gartenzaun ansprechen.""",

    MessageVariant.QUIET_EXPERT: """\
STRATEGIE: "Der stille Experte"
Psychologie: Autorität + Commitment

Ultra-kurze Nachricht. Maximal 35 Wörter. Selbstbewusst durch Kürze. \
Eine scharfe Beobachtung, eine Implikation, eine Ja/Nein-Frage.

STRUKTUR:
1. Fakten-Zusammenfassung in einem Satz (Preis, Fläche, Ort)
2. Eine leicht provozierende Implikation (z.B. "das wird Aufmerksamkeit bekommen, \
aber vermutlich nicht die richtige Art")
3. Einfache Ja/Nein-Frage mit minimaler Reibung

WICHTIG: Maximal 35 Wörter. Jedes Wort muss sitzen.""",

    MessageVariant.VALUE_SPOTTER: """\
STRATEGIE: "Der Wertentdecker"
Psychologie: Reziprozität + Autorität

Gib dem Verkäufer einen echten, nützlichen Insight über sein eigenes Objekt, \
den er vielleicht nicht bedacht hat. Reines Geben ohne zu fragen.

STRUKTUR:
1. Ein verstecktes Wertpotenzial benennen (Einliegerwohnung, Teilbarkeit, \
Mieteinnahmen, Grundstücksnutzung)
2. Kurz erklären, warum das den Wert beeinflusst (1 Satz)
3. Frage, ob sie das bei der Preisfindung berücksichtigt haben

Der Verkäufer soll denken: "Hmm, daran habe ich nicht gedacht" und antworten wollen.""",
}


# --- Follow-up templates ---

FOLLOWUP_1_PROMPT = """\
Du schreibst eine zweite Nachricht an einen Immobilienverkäufer auf Kleinanzeigen, \
der auf deine erste Nachricht nicht geantwortet hat.

REGELN:
- Maximal 50 Wörter
- Gib einen genuinen, nützlichen Markt-Insight (z.B. vergleichbare Verkäufe in der Gegend)
- KEIN Vorwurf, dass nicht geantwortet wurde
- KEIN erneuter Call-to-Action — einfach nur wertvolle Info geben
- Kein "Makler", "Vermittlung", "Provision", "helfen", "unterstützen"
- Beginne mit "Nochmal kurz zum" oder ähnlich beiläufig
- KEINE Grußformel

{tone_instruction}

Informationen:
{listing_context}
"""

FOLLOWUP_2_PROMPT = """\
Du schreibst eine dritte und letzte Nachricht an einen Immobilienverkäufer, der bisher \
nicht geantwortet hat. Diese Nachricht soll ehrlich und abschließend sein.

REGELN:
- Maximal 60 Wörter
- Entschuldige dich kurz für die ungebetenen Nachrichten
- Erwähne beiläufig, dass du "beruflich mit Immobilien in der Region zu tun hast"
- Biete eine "zweite Meinung zum Preis" an, falls sie jemals eine brauchen
- Beende mit "Alles Gute mit dem Verkauf" oder ähnlich
- Das ist die LETZTE Nachricht — sie soll warm und respektvoll sein
- Kein "Makler", "Vermittlung", "Provision"

{tone_instruction}

Informationen:
{listing_context}
"""


def build_listing_context(signals) -> str:
    """Build the context string the LLM receives about the listing."""
    parts = []

    parts.append(f"IMMOBILIE: {signals.property_type}")
    if signals.title:
        parts.append(f"Titel: {signals.title}")
    if signals.city:
        parts.append(f"Ort: {signals.city} ({signals.plz})")
    if signals.price:
        price_str = f"{signals.price:,}€".replace(",", ".")
        if signals.is_vb:
            price_str += " VB"
        parts.append(f"Preis: {price_str}")
    if signals.price_per_sqm:
        parts.append(f"Preis/m²: ~{int(signals.price_per_sqm)}€")
    if signals.wohnflaeche:
        parts.append(f"Wohnfläche: {int(signals.wohnflaeche)}m²")
    if signals.grundstueck:
        parts.append(f"Grundstück: {int(signals.grundstueck)}m²")
    if signals.zimmer:
        parts.append(f"Zimmer: {signals.zimmer}")
    if signals.baujahr:
        parts.append(f"Baujahr: {signals.baujahr}")

    return "\n".join(parts)


def build_personalization_context(signals, personalization) -> str:
    """Build the personalization context string for the LLM."""
    parts = []

    parts.append(f"PERSONALISIERUNG:")
    parts.append(f"Hauptanker (verwende dies als Einstieg): {personalization.primary_anchor}")

    if personalization.secondary_anchors:
        parts.append(f"Weitere Details: {', '.join(personalization.secondary_anchors)}")

    if personalization.price_insight:
        parts.append(f"Preis-Insight: {personalization.price_insight}")

    if personalization.emotional_hook:
        parts.append(f"Emotionaler Aufhänger: {personalization.emotional_hook}")

    # Seller psychology hints
    parts.append(f"\nVERKÄUFER-PSYCHOLOGIE:")
    parts.append(f"Detailgrad der Anzeige: {signals.description_effort.value}")
    parts.append(f"Stimmung: {signals.seller_emotion.value}")
    if signals.is_vb:
        parts.append("Preis ist verhandelbar (VB)")
    if signals.has_provision_note:
        parts.append("Verkäufer erwähnt Provision = ist sich Makler-Dynamik bewusst")

    if signals.renovation_history:
        parts.append(f"Renovierung: {signals.renovation_history}")

    if signals.lifestyle_signals:
        parts.append(f"Lifestyle-Signale: {', '.join(signals.lifestyle_signals[:5])}")

    if signals.unique_features:
        parts.append(f"Besondere Merkmale: {'; '.join(signals.unique_features[:5])}")

    return "\n".join(parts)


def build_generation_prompt(
    signals,
    personalization,
    variant: MessageVariant,
) -> tuple[str, str]:
    """Build the full system prompt and user prompt for LLM generation.

    Returns (system_prompt, user_prompt).
    """
    tone_instruction = TONE_DU if signals.tone.value == "du" else TONE_SIE
    variant_instruction = VARIANT_INSTRUCTIONS[variant]

    system = SYSTEM_PROMPT.format(
        tone_instruction=tone_instruction,
        variant_instruction=variant_instruction,
    )

    user = (
        build_listing_context(signals)
        + "\n\n"
        + build_personalization_context(signals, personalization)
        + "\n\nSchreibe jetzt die Nachricht."
    )

    return system, user


def build_followup_prompt(
    signals,
    stage: int,
) -> tuple[str, str]:
    """Build prompt for follow-up messages.

    stage: 1 for first follow-up, 2 for second (final).
    Returns (system_prompt, user_prompt).
    """
    tone_instruction = TONE_DU if signals.tone.value == "du" else TONE_SIE
    listing_context = build_listing_context(signals)

    if stage == 1:
        system = FOLLOWUP_1_PROMPT.format(
            tone_instruction=tone_instruction,
            listing_context=listing_context,
        )
    else:
        system = FOLLOWUP_2_PROMPT.format(
            tone_instruction=tone_instruction,
            listing_context=listing_context,
        )

    user = "Schreibe jetzt die Nachricht."
    return system, user

File: packages/llm/src/messaging/templates.ts
import {
	IDENTITY,
	STYLE_RULES,
	TASK_INITIAL,
	TASK_FOLLOWUP_1,
	TASK_FOLLOWUP_2,
	injectPersona,
} from "@scraper/agent";
import type { ListingSignals, PersonalizationResult } from "./models";

// --- Persona config ---

export interface MessagePersona {
	name: string;
	firma: string;
	region?: string;
}

const DEFAULT_PERSONA: MessagePersona = {
	name: "Max",
	firma: "Maklermatch",
};

// --- Tone helpers ---

function buildToneDu(vorname: string): string {
	return `ANREDE: Du. Locker, wie mit einem Bekannten. Grußformel: 'VG ${vorname}' oder 'Grüße ${vorname}'.`;
}

function buildToneSie(vorname: string): string {
	return `ANREDE: Sie. Respektvoll aber menschlich, kein Amtsdeutsch. Grußformel: 'Viele Grüße, ${vorname}' oder 'Beste Grüße, ${vorname}'.`;
}

// --- System prompt assembly ---

function buildSystemPrompt(
	task: string,
	toneInstruction: string,
	persona: MessagePersona,
): string {
	const vorname = persona.name.split(" ")[0];
	const parts = [IDENTITY, STYLE_RULES, toneInstruction, task].join("\n\n");
	return injectPersona(parts, vorname, persona.firma);
}

// --- Context builders ---

export function buildListingContext(signals: ListingSignals): string {
	const parts: string[] = [];

	parts.push(`IMMOBILIE: ${signals.propertyType}`);
	if (signals.sellerName)
		parts.push(
			`VERKÄUFER: ${signals.sellerName} (benutze den Vornamen in der Ansprache, nicht den vollen Namen!)`,
		);
	if (signals.title) parts.push(`Titel: ${signals.title}`);
	if (signals.city) parts.push(`Ort: ${signals.city} (${signals.plz})`);
	if (signals.price) {
		let priceStr = `${signals.price.toLocaleString("de-DE")}€`;
		if (signals.isVb) priceStr += " VB";
		parts.push(`Preis: ${priceStr}`);
	}
	if (signals.pricePerSqm)
		parts.push(`Preis/m²: ~${Math.round(signals.pricePerSqm)}€`);
	if (signals.wohnflaeche)
		parts.push(`Wohnfläche: ${Math.round(signals.wohnflaeche)}m²`);
	if (signals.grundstueck)
		parts.push(`Grundstück: ${Math.round(signals.grundstueck)}m²`);
	if (signals.zimmer) parts.push(`Zimmer: ${signals.zimmer}`);
	if (signals.baujahr) parts.push(`Baujahr: ${signals.baujahr}`);

	return parts.join("\n");
}

export function buildPersonalizationContext(
	signals: ListingSignals,
	personalization: PersonalizationResult,
): string {
	const parts: string[] = [];

	parts.push("PERSONALISIERUNG:");
	parts.push(
		`Hauptanker (verwende dies als Einstieg): ${personalization.primaryAnchor}`,
	);

	if (personalization.secondaryAnchors.length > 0)
		parts.push(
			`Weitere Details: ${personalization.secondaryAnchors.join(", ")}`,
		);

	if (personalization.priceInsight)
		parts.push(`Preis-Insight: ${personalization.priceInsight}`);

	if (personalization.emotionalHook)
		parts.push(`Emotionaler Aufhänger: ${personalization.emotionalHook}`);

	parts.push("\nVERKÄUFER-PSYCHOLOGIE:");
	parts.push(`Detailgrad der Anzeige: ${signals.descriptionEffort}`);
	parts.push(`Stimmung: ${signals.sellerEmotion}`);
	if (signals.isVb) parts.push("Preis ist verhandelbar (VB)");
	if (signals.hasProvisionNote)
		parts.push("Verkäufer erwähnt Provision = ist sich Makler-Dynamik bewusst");

	if (signals.renovationHistory)
		parts.push(`Renovierung: ${signals.renovationHistory}`);

	if (signals.lifestyleSignals.length > 0)
		parts.push(
			`Lifestyle-Signale: ${signals.lifestyleSignals.slice(0, 5).join(", ")}`,
		);

	if (signals.uniqueFeatures.length > 0)
		parts.push(
			`Besondere Merkmale: ${signals.uniqueFeatures.slice(0, 5).join("; ")}`,
		);

	return parts.join("\n");
}

// --- Public prompt builders ---

export function buildGenerationPrompt(
	signals: ListingSignals,
	personalization: PersonalizationResult,
	persona: MessagePersona = DEFAULT_PERSONA,
): [string, string] {
	const vorname = persona.name.split(" ")[0];
	const toneInstruction =
		signals.tone === "du" ? buildToneDu(vorname) : buildToneSie(vorname);

	const system = buildSystemPrompt(TASK_INITIAL, toneInstruction, persona);

	const sellerInstruction = signals.sellerName
		? `\n\nDer Verkäufer heißt ${signals.sellerName}. Benutze nur den Vornamen in der Ansprache.`
		: "";

	const user =
		buildListingContext(signals) +
		"\n\n" +
		buildPersonalizationContext(signals, personalization) +
		sellerInstruction +
		"\n\nSchreibe jetzt die Nachricht.";

	return [system, user];
}

export function buildFollowupPrompt(
	signals: ListingSignals,
	stage: number,
	persona: MessagePersona = DEFAULT_PERSONA,
): [string, string] {
	const vorname = persona.name.split(" ")[0];
	const toneInstruction =
		signals.tone === "du" ? buildToneDu(vorname) : buildToneSie(vorname);

	const task = stage === 1 ? TASK_FOLLOWUP_1 : TASK_FOLLOWUP_2;
	const system = buildSystemPrompt(task, toneInstruction, persona);

	const user =
		buildListingContext(signals) + "\n\nSchreibe jetzt die Nachricht.";

	return [system, user];
}
